File: packages/silo-client/src/media/media-delete-failure.ts
import type { MediaUsage } from "./media-usage.js";

/**
 * One id's outcome in a bulk delete that could not remove it. `usageCount`,
 * `visibleCount`, `visibleCapped` and `referrers` are present only for the
 * `"media_in_use"` code — `"not_found"`, `"media_delete_stalled"` and
 * `"invalid_id"` carry just `id`, `code` and `message`.
 */
export interface MediaDeleteFailure {
  id: string;
  code: string;
  message: string;
  usageCount?: number;
  visibleCount?: number;
  visibleCapped?: boolean;
  referrers?: MediaUsage[];
}

File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/media/MediaDeleteFailure.java
package in.org.quicko.silo.client.media;

import com.fasterxml.jackson.databind.JsonNode;
import in.org.quicko.silo.client.transport.JsonValues;
import java.util.List;

/**
 * One id's outcome in a bulk delete that could not remove it.
 *
 * <p>The counts and the referrers are present only for the
 * {@code "media_in_use"} code — {@code "not_found"},
 * {@code "media_delete_stalled"} and {@code "invalid_id"} carry just the id, the
 * code and the message.
 */
public record MediaDeleteFailure(
    String id,
    String code,
    String message,
    long usageCount,
    long visibleCount,
    boolean visibleCapped,
    List<MediaUsage> referrers) {

  public MediaDeleteFailure {
    referrers = List.copyOf(referrers);
  }

  public static MediaDeleteFailure fromWire(JsonNode payload) {
    return new MediaDeleteFailure(
        JsonValues.text(payload, "id"),
        JsonValues.text(payload, "code"),
        JsonValues.text(payload, "message"),
        payload.path("usage_count").asLong(0),
        payload.path("visible_count").asLong(0),
        payload.path("visible_capped").asBoolean(false),
        MediaMapper.toUsages(payload.get("referrers")));
  }
}
